app/assets/javascripts/components/ecnumbers.js

```diff
@@ -0,0 +1,260 @@
+        this.ec = new this[addMissing](data);
+        ECNumbers[addNames](data);
+
+        // Fetch names in the background, not needed yet
+        setTimeout(()=>{
+            ECNumbers[addMissingNames](Array.from(this.ec.keys()));
+        }, 0);
+    }
+
+    /**
+     * Returns the originality supplied set of EC numbers
+     * sorted by value
+     * @return {[FACounts]} EC number values
+     */
+    sortedTerms() {
+        return this.data;
+    }
+
+    /**
+     * Create a map width all EC numbers given and their ancestors even if they
+     * are not given.
+     * @param {[FACounts]} newEC list of new EC data
+     * @param {Map} map map to start form
+     * @return {Map}
+     * @access private
+     */
+    [addMissing](newEC, map = null) {
+        let result = (map === null ? new Map(newEC.map(x=>[x.code, x])) : map);
+
+        for (let curEc of newEC) {
+            const parts = curEc.code.split(".");
+            const numSpecific = parts.includes("-") ? parts.indexOf("-") : parts.length;
+            result.set(curEc.code, curEc); // overrides if exists
+
+            for (let i = numSpecific-1; i>=1; i--) {
+                parts[i] = "-";
+                const newKey = parts.join(".");
+                if (!result.has(newKey)) {
+                    let parentEC = {code: newKey, name: null, value: 0};
+                    parentEC[artificial] = true;
+                    result.set(newKey, parentEC);
+                } else {
+                    break;// the key already exists (all following already done)
+                }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Gets the value of the EC number
+     *
+     * @param  {string} ecNum The code of the EC number (like "2.3.-.-")
+     * @return {number}       The value of the EC number
+     */
+    getValueOf(ecNum) {
+        return this.ec.get(ecNum).value || 0;
+    }
+
+    /**
+     * Gets the value of the EC number as fraction of the number of
+     * the number of annotated peptides
+     *
+     * @param  {string} ecNum The code of the EC number (like "2.3.-.-")
+     * @return {string}       The name of the EC number
+     */
+    getFractionOf(ecNum) {
+        return this.getValueOf(ecNum) / this.numTotalSet;
+    }
+
+    /**
+     * Number of annotated peptides
+     *
+     * @return {number}       Number of annotated peptides
+     */
+    getTotalSetSize() {
+        return this.numTotalSet;
+    }
+
+    /**
+     * Create an EC tree
+     *
+     * The numbers are first sorted form general to specific (number of - at the
+     * end). Then the tree is built by adding the count to all ancestors.
+     *
+     * @param {string} target the element to put the tree in (will be cleared)
+     * @param {object} treeViewOptions options to pass to the treeview
+     * @return {TreeView} the created tree
+     */
+    createTree(target, treeViewOptions) {
```

Comment:
This function is quite long, consider splitting it up if possible.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13720) by @bmesuere on Mon Feb 19 2018 at 14:03._